src/tests/overlay_kv.rs

```diff
@@ -0,0 +1,110 @@
+use std::collections::BTreeMap;
+
+use proptest::{collection::btree_map, prelude::*};
+use tokio::{runtime::Runtime, sync::RwLock};
+
+use crate::{
+    mock::MockTreeStore,
+    restore::{JellyfishMerkleRestore, StateSnapshotReceiver},
+    storage::TreeReader,
+    tests::helper::init_mock_db,
+    JellyfishMerkleTree, KeyHash, OwnedValue, RootHash, Version, WriteOverlay,
+};
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum Action {
+    Put { key: KeyHash, value: OwnedValue },
+    Get { key: KeyHash },
+    Commit,
+}
+
+#[derive(Clone, Debug, Default)]
+struct MockKvStore {
+    store: BTreeMap<KeyHash, OwnedValue>,
+}
+
+impl Action {
+    async fn run(
+        &self,
+        overlay: &mut WriteOverlay<MockTreeStore>,
+        mock_tree_store: MockTreeStore,
+        mock_kv: &mut MockKvStore,
+    ) -> anyhow::Result<()> {
+        match self {
+            Action::Put { key, value } => {
+                overlay.put(*key, value.clone());
+                mock_kv.store.insert(*key, value.clone());
+            }
+            Action::Get { key } => {
+                let overlay_value = overlay.get(*key).await?;
+                let mock_kv_value = mock_kv.store.get(key);
+                assert_eq!(
+                    overlay_value.as_ref(),
+                    mock_kv_value,
+                    "`get` returned different values"
+                );
+            }
+            Action::Commit => {
+                overlay.commit(mock_tree_store).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn empty_commit() {
+    let mock_tree_store = MockTreeStore::default();
+    let mut overlay = WriteOverlay::new(mock_tree_store.clone(), 0);
+    overlay.commit(mock_tree_store).await.unwrap();
+}
+
+#[tokio::test]
+async fn put_then_commit() {
+    let mock_tree_store = MockTreeStore::default();
+    let mut overlay = WriteOverlay::new(mock_tree_store.clone(), 0);
+    overlay.put(b"".into(), b"".to_vec());
+    overlay.commit(mock_tree_store).await.unwrap();
+}
```

Comment:
I don't understand why this test is failing, or why this issue is not being triggered when we use the `WriteOverlay` in production.